File: apps/api-java/src/main/java/com/nanaskitchens/api/auth/AuthService.java
package com.nanaskitchens.api.auth;

import com.nanaskitchens.api.audit.AuditLog;
import com.nanaskitchens.api.audit.AuditLogRepository;
import com.nanaskitchens.api.auth.dto.TokenPairResponse;
import com.nanaskitchens.api.common.Role;
import com.nanaskitchens.api.security.JwtService;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Base64;
import java.util.EnumSet;
import java.util.HexFormat;
import java.util.Set;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Service
public class AuthService {

    private static final Set<Role> SELF_SERVE_ROLES = EnumSet.of(Role.buyer, Role.seller);
    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final AuditLogRepository auditLogRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final long refreshTokenTtlDays;

    public AuthService(
            UserRepository userRepository,
            RefreshTokenRepository refreshTokenRepository,
            AuditLogRepository auditLogRepository,
            PasswordEncoder passwordEncoder,
            JwtService jwtService,
            @Value("${app.jwt.refresh-token-ttl-days:30}") long refreshTokenTtlDays) {
        this.userRepository = userRepository;
        this.refreshTokenRepository = refreshTokenRepository;
        this.auditLogRepository = auditLogRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
        this.refreshTokenTtlDays = refreshTokenTtlDays;
    }

    @Transactional
    public TokenPairResponse register(String email, String rawPassword, String roleValue) {
        Role role = parseSelfServeRole(roleValue);
        if (userRepository.existsByEmail(email)) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "EMAIL_TAKEN");
        }
        User user = new User();
        user.setEmail(email);
        user.setPasswordHash(passwordEncoder.encode(rawPassword));
        user.setRole(role);
        user = userRepository.save(user);
        audit(user.getId(), "User", "register");
        return issuePair(user);
    }

    @Transactional
    public TokenPairResponse login(String email, String rawPassword) {
        User user = userRepository
                .findByEmail(email)
                .filter(u -> passwordEncoder.matches(rawPassword, u.getPasswordHash()))
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "INVALID_CREDENTIALS"));
        audit(user.getId(), "User", "login");
        return issuePair(user);
    }

    /** Refresh rotation: previous token is single-use and revoked once redeemed (mirrors AC3). */
    @Transactional
    public TokenPairResponse refresh(String rawRefreshToken) {
        String tokenHash = sha256Hex(rawRefreshToken);
        LocalDateTime now = LocalDateTime.now(ZoneOffset.UTC);
        RefreshToken row = refreshTokenRepository
                .findByTokenHash(tokenHash)
                .filter(rt -> rt.getRevokedAt() == null && rt.getExpiresAt().isAfter(now))
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "REFRESH_INVALID"));
        row.setRevokedAt(now);
        refreshTokenRepository.save(row);

        User user = userRepository
                .findById(row.getUserId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "REFRESH_INVALID"));
        return issuePair(user);
    }

    private TokenPairResponse issuePair(User user) {
        String accessToken = jwtService.issueAccessToken(user.getId(), user.getRole());
        String rawRefreshToken = generateOpaqueToken();

        RefreshToken refreshToken = new RefreshToken();
        refreshToken.setUserId(user.getId());
        refreshToken.setTokenHash(sha256Hex(rawRefreshToken));
        refreshToken.setExpiresAt(LocalDateTime.now(ZoneOffset.UTC).plusDays(refreshTokenTtlDays));
        refreshTokenRepository.save(refreshToken);

        return new TokenPairResponse(accessToken, rawRefreshToken);
    }

    private Role parseSelfServeRole(String raw) {
        Role role;
        try {
            role = Role.valueOf(raw);
        } catch (IllegalArgumentException e) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "UNKNOWN_ROLE");
        }
        if (!SELF_SERVE_ROLES.contains(role)) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "ROLE_NOT_SELF_SERVE");
        }
        return role;
    }

    private static String generateOpaqueToken() {
        byte[] bytes = new byte[48];
        SECURE_RANDOM.nextBytes(bytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }

    private static String sha256Hex(String value) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return HexFormat.of().formatHex(digest.digest(value.getBytes(StandardCharsets.UTF_8)));
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
    }

    private void audit(String actor, String entity, String action) {
        AuditLog log = new AuditLog();
        log.setActor(actor);
        log.setEntity(entity);
        log.setAction(action);
        auditLogRepository.save(log); // NFR10
    }
}

File: apps/api/src/auth/auth.service.ts
import { ForbiddenException, Injectable, UnauthorizedException } from "@nestjs/common";
import { JwtService } from "@nestjs/jwt";
import * as argon2 from "argon2";
import { createHash, randomBytes } from "node:crypto";
import { PrismaService } from "../prisma/prisma.service";
import type { Role } from "@culture-eats/core";

const SELF_SERVE_ROLES: Role[] = ["buyer", "seller"]; // inspector/admin yalnizca admin tarafindan atanir (AC2)

@Injectable()
export class AuthService {
  constructor(private prisma: PrismaService, private jwt: JwtService) {}

  async register(email: string, password: string, role: Role) {
    if (!SELF_SERVE_ROLES.includes(role)) {
      throw new ForbiddenException("ROLE_NOT_SELF_SERVE");
    }
    const passwordHash = await argon2.hash(password);
    const user = await this.prisma.user.create({ data: { email, passwordHash, role } });
    await this.audit(user.id, "User", "register");
    return this.issuePair(user.id, user.role);
  }

  async login(email: string, password: string) {
    const user = await this.prisma.user.findUnique({ where: { email } });
    if (!user || !(await argon2.verify(user.passwordHash, password))) {
      throw new UnauthorizedException("INVALID_CREDENTIALS");
    }
    await this.audit(user.id, "User", "login");
    return this.issuePair(user.id, user.role);
  }

  /** Refresh rotasyonu: eski token tek kullanimlik, dogrulaninca iptal edilir (AC3). */
  async refresh(refreshToken: string) {
    const tokenHash = this.hash(refreshToken);
    const row = await this.prisma.refreshToken.findUnique({ where: { tokenHash }, include: { user: true } });
    if (!row || row.revokedAt || row.expiresAt < new Date()) {
      throw new UnauthorizedException("REFRESH_INVALID");
    }
    await this.prisma.refreshToken.update({ where: { id: row.id }, data: { revokedAt: new Date() } });
    return this.issuePair(row.user.id, row.user.role);
  }

  private async issuePair(sub: string, role: string) {
    const accessToken = await this.jwt.signAsync({ sub, role });
    const refreshToken = randomBytes(48).toString("base64url");
    await this.prisma.refreshToken.create({
      data: { userId: sub, tokenHash: this.hash(refreshToken), expiresAt: new Date(Date.now() + 30 * 864e5) },
    });
    return { accessToken, refreshToken };
  }

  private hash(t: string) { return createHash("sha256").update(t).digest("hex"); }

  private audit(actor: string, entity: string, action: string) {
    return this.prisma.auditLog.create({ data: { actor, entity, action } }); // NFR10
  }
}
